guide/view-page-search-parameters.md
---
outline: deep
---

# 테이블 검색 조건 추가

데이터를 조회할 때, 특정 조건과 값으로 필요한 정보를 빠르게 찾을 수 있는 페이지가 필요할 때가 있습니다. 셀렉트 어드민을 통해 복잡한 검색 조건 기능을 쉽게 만들어보세요. 

> 📘 
> 
> 본 문서는 mysql 을 기준으로 작성되었습니다.  
> postgresql을 이용하시는 경우 ["검색 조건 PostgreSQL 유의사항"](/guide/tip-postgresql)을 참고해주시기 바랍니다.

## 필수 검색 조건

데이터를 조회할 때, 입력 필드를 꼭 채워야하는 방법입니다.

![](https://imagedelivery.net/MHVC-FGTDyxApYeHyF29Tw/bd3ba0ab-a288-4618-c722-2e2bb4574800/docs "necessary-filter.png")

아래 쿼리에서 `:name` 값에 아무것도 들어가지 않으면 `''` 빈값으로 쿼리가 실행됩니다. 

```yaml
- path: properties/active
  blocks:
  - type: query
    resource: mysql.qa
    sqlType: select
    sql: >
      SELECT * 
      FROM properties 
      WHERE name = :name
      ORDER BY id ASC 
      LIMIT 100
    # 검색조건 설정 
    params:
    - key: name
```

## 필수 조건 여러개

여러개의 필드를 조건으로 검색하고 싶을 때 

![](https://imagedelivery.net/MHVC-FGTDyxApYeHyF29Tw/3275fc12-5ab6-4e61-cc82-b47ed41e1600/docs "multi-filters.png")

입력 필드가 1개 있을 때와 마찬가지로 아래와 같이 작성 시, 파라미터의 값이 없을 때는 빈값 `''`으로 들어가 쿼리가 실행됩니다.  
쿼리의 OR절과 함께 사용하는 경우, 둘중 하나만 값이 맞아도 조회할 수 있게 됩니다.

```yaml
- path: properties/active
  blocks:
  - type: query
    resource: mysql.qa
    sqlType: select
    sql: >
      SELECT * 
      FROM properties 
      WHERE name = :name
         OR id = :id
      ORDER BY id ASC 
      LIMIT 100
    params:
    - key: name
      label: 업체명
    - key: id
      label: 업체번호
      format: number
```

## 조건을 입력해야 검색하기

기본은 전체 보기, 조건 둘중 하나라도 해당되면 표시

```yaml
- path: properties/active
  blocks:
  - type: query
    resource: mysql.qa
    sqlType: select
    sql: >
      SELECT * 
      FROM properties 
      WHERE (LENGTH(:name) = 0 OR name = :name)
         OR (LENGTH(:id) = 0 OR id = :id)
      ORDER BY id ASC 
      LIMIT 100
    params:
    - key: name
      label: 업체명
    - key: id
      label: 업체번호
      format: number
```

## 기본은 모두 표시, 조건 모두 일치해야 표시

```yaml
- path: properties/active
  blocks:
  - type: query
    resource: mysql.qa
    sqlType: select
    sql: >
      SELECT * 
      FROM properties 
      WHERE 1=1 
        AND (LENGTH(:name) = 0 OR name = :name)
        AND (LENGTH(:id) = 0 OR id = :id)
      ORDER BY id ASC 
      LIMIT 100
    params:
    - key: name
      label: 업체명
    - key: id
      label: 업체번호
      format: number
```

## 드롭다운 dropdown으로 조건 검색

![](https://imagedelivery.net/MHVC-FGTDyxApYeHyF29Tw/347d27bf-48e0-4f44-b9aa-b0db761b0000/docs "dropdown-search.png")

```yaml
- type: query
  resource: mysql
  sqlType: select
  sql: >
    SELECT *
    FROM orders
    WHERE (LENGTH(:status) = 0 OR status = :status)
  params:
  - key: status
    label: 상태
    dropdown:
    - ''
    - in_bucket
    - paid
    - in_delivery
    - delivered
    - confirmed
```

## 자동완성 auto-complete 설정

### datalist

코드와 레이블 모두를 검색합니다.

![](https://imagedelivery.net/MHVC-FGTDyxApYeHyF29Tw/c09b6142-a4f0-49fd-9086-bfca015bfe00/docs "autocomplete-search.png")

```yaml
- path: properties/active
  blocks:
  - type: query
    resource: mysql.qa
    sqlType: select
    sql: >
      SELECT * 
      FROM properties 
      WHERE (LENGTH(:type) = 0 OR `type` = :type)
      ORDER BY id ASC 
      LIMIT 100
    params:
    - key: type
      label: 분류코드
      # 자동완성 설정
      datalist: 
      - value: A000
        label: 분류1
      - value: A002
        label: 분류2
```

### 데이터를 쿼리로 가져오기

![](https://imagedelivery.net/MHVC-FGTDyxApYeHyF29Tw/4d6c65d6-52ee-4389-d27d-023e0bf6e300/docs "datalist-from-query.png")

```yaml
- path: properties/active
  blocks:
  - type: query
    resource: mysql.qa
    sqlType: select
    sql: >
      SELECT * 
      FROM properties 
      WHERE (LENGTH(:type) = 0 OR `type` = :type)
      ORDER BY id ASC 
      LIMIT 100
    params:
    - key: type
      label: 분류코드
      datalist: true
      # 자동완성에 쓰는 데이터를 쿼리로 가져오기
      datalistFromQuery: 
        type: query
        resource: mysql.qa
        sql: SELECT DISTINCT `type` FROM properties
```

### 라디오 박스 radio box로 조건 검색

![](https://imagedelivery.net/MHVC-FGTDyxApYeHyF29Tw/43968a47-cf3c-4c02-7d35-9b85b1ce1e00/docs "radiobox-search.png")

```yaml
- path: orders-status
  blocks:
  - type: query
    resource: pgsql
    sqlType: select
    sql: >
      SELECT *
      FROM orders
      WHERE (LENGTH(:status) = 0 OR status = :status)
    params:
    - key: status
      label: 상태
      radio:
      - value: ''
        label: 전체
      - value: in_bucket
        label: 장바구니
      - value: paid
        label: 결제완료
      - value: in_delivery
        label: 배송중
      - value: delivered
        label: 배송완료
      - value: confirmed
        label: 구매확정
```

### 기본값 설정

```yaml
params:
- key: status
  label: 상태
  # 기본값 설정
  defaultValue: ALL
  radio:
  - value: ALL
    label: 전체
```

### 체크박스 checkbox로 조건 검색

**예제1**

![](https://imagedelivery.net/MHVC-FGTDyxApYeHyF29Tw/acd9bfe3-7f94-449d-2e7c-ccdfa38de400/docs "스크린샷 2023-01-06 오후 10.18.20.png")

```yaml
pages:
- path: featured-products
  blocks:
  - type: query
    resource: mysql.qa
    sqlType: select
    sql: > 
      SELECT * FROM products
      WHERE (LENGTH(:is_featured) = 0 OR is_featured = :is_featured) 
      ORDER BY id DESC LIMIT 100
    params:
    - key: is_featured
      label: 피처드
      checkbox:
        true: 1
        false: 0
      type: number
    columnOptions:
    - field: id
    - field: name
    - field: is_featured
```

**예제2**

checkbox와 {{query}}를 같이 사용할 때

![](https://imagedelivery.net/MHVC-FGTDyxApYeHyF29Tw/9dd5f012-3333-416e-3daa-48bb48032300/docs "checkbox-search.png")

true는 체크박스가 체크되었을 때의 값이고, false는 체크 해제되었을 때의 값입니다.

```yaml
pages:
- path: users
	blocks:
  - type: query
  	resource: mysql.qa
    sqlType: select
    sql: >
    	SELECT *
      FROM users
      WHERE {{query}}
      LIMIT 100
    params:
    - key: filter
      checkbox:
        true: WITHOUT
        false: WITH
        label: 사용자만
      defaultValue: WITHOUT
      query:
        WITH: 1=1
        WITHOUT: id NOT IN (1, 2, 3)
```

## 검색 조건에 날짜 기본값 넣기

데이터 조회 시 당일 또는 특정 기준의 날짜를 기본값으로 지정하고 싶을 때 유용합니다.

### defaultValueFromQuery

쿼리로 날짜 데이터 가져와서 기본값으로 넣는 방법입니다. 페이지 조회 시 1회 실행되며, sql 쿼리 결과의 첫번째 값 또는 AS value 값을 가져옵니다. 

```yaml
params:
- key: created_at
  defaultValueFromQuery:
    type: query
    resource: acme
    sql: SELECT date_format(now(), '%Y-%m-%d')
```

### defaultValueFn

자바스크립트(Javascript, JS)로 데이터를 가져와서 기본값으로 넣는 방법입니다. 

> 👍 
> 
> 현재 [moment](https://momentjs.com/), [Luxon](https://moment.github.io/luxon/#/?id=luxon) 을 기본으로 지원하고 있습니다.

```yaml
params:
- key: created_at
  defaultValueFn: |
    // return moment().add(3, 'days').format('YYYY-MM-DD')
    // return DateTime.now().toFormat('yyyy-MM-dd')
    return '2022-08-12'
```

> 📘 yaml 여러줄 입력 시 > 와 | 의 차이점
> 
> yaml에서 여러줄을 입력할 때는 `>` 와 `|` 모두 가능하지만 중요한 차이점이 있습니다. 
> 
> - `>` 는 줄바꿈 시 trim(문자열 양 끝의 공백이 제거)되기 때문에 줄 간 구분을 하기 위해서는 `;` 를 잘 써야합니다. 
> - `|` 는 trim 되지 않습니다.

.vitepress/config.mjs
import { defineConfig } from 'vitepress'

// https://vitepress.dev/reference/site-config
export default defineConfig({
  title: "selectfromuser.com",
  description: "Select Admin",
  cleanUrls: true,
  markdown: {
    lineNumbers: true,
  },
  head: [
    [
      'script',
      {
        defer: '',
        'data-domain': 'dev.selectfromuser.com',
        src: 'https://plausible.io/js/script.js',
      }
    ]
  ],
  themeConfig: {
    // https://vitepress.dev/reference/default-theme-config
    nav: [
      { text: 'Home', link: '/' },
      { text: 'Guide', link: '/guide/' },
      { text: 'Reference', link: '/layout' },
      { text: 'Changelog', link: '/changelog/' },
      { text: 'Community', link: 'https://ask.selectfromuser.com' },
    ],

    sidebar: {
      '/guide/': [
        // {
        //   items: [
        //   ]
        // },
        {
          text: '시작하기',
          items: [
            { text: '셀렉트란', link: '/guide/'},
            { text: 'Quickstart', link: '/guide/quickstart'},            
          ]
        },
        {
          text: '어드민 만들기',
          items: [
            { 
              text: '어드민 편집', 
              link: '/guide/admin-edit', 
              collapsed: true,
              items: [
                { text: '편집 결과 미리보기', link: '/guide/admin-edit-preview'},
              ],
            },
            { 
              text: '메뉴 구성', 
              link: '/guide/menu', 
              collapsed: true,
              items: [
                { text: '메뉴 권한 설정', link: '/guide/menu-permission'},
                { text: '메뉴 path 이름 규칙', link: '/guide/menu-naming-path'},
                { text: '계층 메뉴 만들기', link: '/guide/menu-submenu-tree'},
              ],
            },
            { 
              text: '데이터 조회 페이지', 
              link: '/guide/view-page', 
              collapsed: true,
              items: [
                { text: '테이블 페이지네이션', link: '/guide/view-page-pagination' },
                { text: '테이블 정렬 방법', link: '/guide/view-page-order' },
                { text: '테이블 필드 이름 바꾸기', link: '/guide/view-page-field' },
                { text: '테이블 간편 검색', link: '/guide/view-page-search' },
                { text: '테이블 검색 조건 추가', link: '/guide/view-page-search-parameters' },
                { text: '테이블 필드 포맷 설정', link: '/guide/view-page-column' },
                { text: '차트, 그래프로 표시', link: '/guide/view-page-graph-chart' },
                { text: '페이지 이동', link: '/guide/view-page-reference' },
                { text: '칸반 보드로 보기', link: '/guide/view-page-board' },
              ],
            },
            { 
              text: '데이터 추가 및 수정 페이지', 
              link: '/guide/crud-page', 
              collapsed: true,
              items: [
                { text: '클릭해서 수정하기', link: '/guide/crud-page-inline-edit' },
                { text: '복수의 데이터 추가, 수정', link: '/guide/crud-page-selected' },
                { text: '텍스트 에디터로 추가, 수정하기', link: '/guide/crud-page-editor' },
                { text: '이미지 파일 추가', link: '/guide/crud-page-image-file' },
              ],
            },
            { 
              text: '설명 추가하기', 
              link: '/guide/describe-page', 
            },
            { 
              text: '입력 필드 만들기', 
              link: '/guide/input', 
              collapsed: true,
              items: [
                { text: '입력 필드에 기본값 넣기', link: '/guide/input-default' },
                { text: '입력 필드 상세 옵션', link: '/guide/input-option' },
                { text: '파일 포맷 업로드', link: '/guide/input-file-upload' },
              ],
            },
            { 
              text: '모달 구성', 
              link: '/guide/modal', 
              collapsed: true,
              items: [
                { text: '모달 상세 옵션', link: '/guide/modal-option' },
              ],
            },
            { 
              text: '데이터 체크와 액션', 
              link: '/guide/action', 
              collapsed: true,
              items: [
                { text: '액션 상세 설정', link: '/guide/action-customize' },
                { text: '입력 필드 상세 옵션', link: '/guide/action-button-open-modal' },
                { text: '여러번 반복해서 실행', link: '/guide/action-for-each' },
              ],
            },
            { 
              text: '모달과 테이블 버튼', 
              link: '/guide/table-buttons-multiple-modal', 
            },
            { 
              text: 'API 활용', 
              link: '/guide/api', 
              collapsed: true,
              items: [
                { text: 'API로 수정, 삭제', link: '/guide/api-crud' },
                { text: 'JSON body 넣는 법', link: '/guide/api-json-body' },
                { text: 'API KEY 안전하게 이용', link: '/guide/api-key' },
                { text: 'GraphQL 사용', link: '/guide/api-graphql' },
                { text: 'API 응답을 코드로 처리', link: '/guide/api-response' },
              ],
            },
            { 
              text: '대시보드 넣기', 
              link: '/guide/dashboard', 
            },
            { 
              text: '구글 스프레드시트 이용', 
              link: '/guide/google-spreadsheet', 
            },
            { 
              text: '여러 데이터베이스 JOIN', 
              link: '/guide/sql-with-database-join', 
            },
            { 
              text: '레이아웃 커스텀', 
              link: '/guide/custom-layout', 
            },
          ]
        },
        {
          text: '샘플',
          items: [
            { 
              text: '고객 지원 툴', 
              link: '/guide/sample-cs-admin', 
            },
            { 
              text: '네이버 지도 API 호출', 
              link: '/guide/sample-naver-maps-api', 
            },
            { 
              text: '콘텐츠 관리 CMS', 
              link: '/guide/sample-content-management-system', 
            },
            { 
              text: '슬랙에 메시지 보내기', 
              link: '/guide/sample-send-to-slack', 
            },
            { 
              text: '상품 재고 관리', 
              link: '/guide/sample-inventory-management', 
            },
          ]
        },
        {
          text: 'SQL 팁과 유의사항',
          items: [
            { 
              text: 'SQL 조건문 팁', 
              link: '/guide/tip-sql', 
            },
            { 
              text: 'PostgreSQL 유의사항', 
              link: '/guide/tip-postgresql', 
            },
          ]
        },
        {
          text: '리소스 연결',
          items: [
            { 
              text: '데이터베이스 연결', 
              link: '/guide/database-connect', 
            },
            { 
              text: '데이터베이스 변경', 
              link: '/guide/database-modify', 
            },
            { 
              text: 'MongoDB', 
              link: '/guide/database-mongodb', 
            },
            { 
              text: 'JSON 데이터 사용', 
              link: '/guide/database-json', 
            },
          ]
        },
        {
          text: '팀 관리하기',
          items: [
            { 
              text: '팀 만들기', 
              link: '/guide/team-new', 
            },
            { 
              text: '팀에 멤버 초대하기', 
              link: '/guide/team-invitation', 
            },
            { 
              text: '사용자 속성 설정', 
              link: '/guide/team-user-property', 
            },
            { 
              text: '팀 설정', 
              link: '/guide/team-setting', 
            },
          ]
        },
        {
          text: '로그 및 보안',
          items: [
            { 
              text: '로깅', 
              link: '/guide/logging', 
            },
            { 
              text: 'SQL 쿼리 보안', 
              link: '/guide/secure-sql-query', 
            },
            { 
              text: 'API 보안', 
              link: '/guide/secure-api', 
            },
            { 
              text: 'SSO, IP, 세션 설정', 
              link: 'https://blog.selectfromuser.com/network-security-policy/', 
            },
          ]
        },
        {
          text: '컴포넌트',
          items: [
            { 
              text: '컴포넌트 라이브러리', 
              link: '/guide/component-library', 
            },
          ]
        },
      ],
      '/changelog': [
        {
          items: [
            { text: 'CHANGELOG.md', link: '/changelog/'},
            { text: '2023-08-03', link: '/changelog/2023-08-03' },
            { text: '2023-07-07', link: '/changelog/2023-07-07' },
            { text: '2023-06-08', link: '/changelog/2023-06-08' },
            { text: '2023-03-30', link: '/changelog/2023-03-30' },
            { text: '2023-03-06', link: '/changelog/2023-03-06' },
            { text: '2023-02-28', link: '/changelog/2023-02-28' },
            { text: '2023-01-08', link: '/changelog/2023-01-08' },
            { text: '2022-11-22', link: '/changelog/2022-11-22' },
            { text: '2022-10-28', link: '/changelog/2022-10-28' },
            { text: '2022-09-30', link: '/changelog/2022-09-30' },
            { text: '2022-08-31', link: '/changelog/2022-08-31' },
            { text: '2022-07-29', link: '/changelog/2022-07-29' },
            { text: '2022-06-30', link: '/changelog/2022-06-30' },
          ]
        },
      ],
      '/': [
        {
          text: 'Overview',
          items: [
            // { text: '시작하기', link: '/start' },
            { text: '개념 및 설계 구조', link: '/introduction' },
          ]
        },
        {
          text: 'Reference',
          items: [
            { text: 'Layout', link: '/layout' },
            { text: 'Menus', link: '/menus' },
            { text: 'Pages', link: '/pages' },
            { text: 'Blocks', link: '/blocks' },
            { text: 'Columns', link: '/columns' },
            { text: 'Params', link: '/params' },
            { text: 'Modals', link: '/modals' },
            { text: 'Actions', link: '/actions' },
          ]
        },
        {
          text: 'Links',
          items: [
            { text: '샘플 예제', link: 'https://showroom.selectfromuser.com' },
            { text: '로그인', link: 'https://app.selectfromuser.com' },
            { text: '(과거) YAML reference', link: '/docs/yaml-reference' },
          ]
        },
      ]
    },
    // sidebar: [
    // ],

    socialLinks: [
      { icon: 'github', link: 'https://github.com/selectfromusercom/docs' }
    ], 

    search: {
      // provider: 'local',
      provider: 'algolia',
      options: {
        appId: process.env.ALGOLIA_APP_ID,
        apiKey: process.env.ALGOLIA_API_KEY,
        indexName: process.env.ALGOLIA_INDEX_NAME,
      }
    },
    editLink: {
      pattern: 'https://github.com/selectfromusercom/docs/edit/main/:path'
    },
  }
})
